File: javascript/7kyu/Driving Licence.js
// https://www.codewars.com/kata/586a1af1c66d18ad81000134

// The UK driving number is made up from the personal details of the driver. The individual letters and digits can be code using the below information

function driver(data) {
    var name = (data[2].slice(0,5).toUpperCase() +'99999').slice(0,5);    // 5
    var dob = new Date(data[3]);
    var dec = (dob.getYear().toString().length==2 ? dob.getYear().toString()[0] : dob.getYear().toString()[1]);  // 1
    var month = (data[4] == "M" ? ("0" + (dob.getMonth()+1).toString()).slice(-2) : dob.getMonth()+51);
    var day = ("0" + dob.getDate().toString()).slice(-2);
    var yearDig = dob.getYear().toString().slice(-1);
    var inits = data[0][0] + (data[1]=="" ? "9" : data[1][0])
    var arb1 = "9"
    var arb2 = "AA"
    return name + dec + month + day + yearDig + inits + arb1 + arb2
  }

File: java/7kyu/Driving Licence.java
// https://www.codewars.com/kata/586a1af1c66d18ad81000134

// The UK driving number is made up from the personal details of the driver. The individual letters and digits can be code using the below information

public class Solution {
	public static final String standardValues = "9AA";
	
	  public static String driver(final String[] data) {
		  
		  String surname = data[2];
		  String[] birthDate = data[3].split("-");
		  String first = data[0];
		  String middle = data[1];
		  String gender = data[4];
		  
		  StringBuilder builder = new StringBuilder();
		  
		  builder.append(formatSurname(surname).toUpperCase());
		  builder.append(birthDate[2].substring(2, 3));
		  builder.append(formatMonthValue(birthDate, gender));
		  builder.append(formatDayValue(birthDate));
		  builder.append(birthDate[2].substring(3, 4));
		  builder.append(formatInitialRule(first));
		  builder.append(formatInitialRule(middle));
		  builder.append(standardValues);
	  
		  return builder.toString().trim(); 
		  }

	private static String formatInitialRule(String name) {
		if(null == name || name.isEmpty()) {
			return "9";
		}

		return name.substring(0, 1);
	}

	private static String formatDayValue(String[] birthDate) {
		String fred = birthDate[0];
		
		return fred.length() == 1? "0".concat(fred) : fred;

	}

	private static String formatMonthValue(String[] birthDate, String gender) {
		String birthMonth = birthDate[1].substring(0, 3).toUpperCase();
		String fred = Months.valueOf(birthMonth).getMonth();
		
		if(gender.toUpperCase().equals("F")) {
			int value = Integer.parseInt(fred.substring(0,1));
			value = value + 5;
			return String.valueOf(value).concat(fred.substring(1,2));
		}
		
		return fred;
	}

	private static String formatSurname(String surname) {
		int length = surname.length();
		StringBuilder builder = new StringBuilder();
		
		if(length >= 5) {
			builder.append(surname.substring(0, 5));
		} else {
			builder.append(surname);
			
			for (int i = 1; i <= 5 - length; i++) {
				builder.append("9");
			}
		}

		return builder.toString().trim();
	}
}

	 enum Months {
		JAN ("01"),
		FEB ("02"),
		MAR ("03"),
		APR ("04"),
		MAY ("05"),
		JUN ("06"),
		JUL ("07"),
		AUG ("08"),
		SEP ("09"),
		OCT ("10"),
		NOV ("11"),
		DEC ("12");
		 
     private String month;
		 
     public String getMonth()
	   {
	      return this.month;
       }

	 private Months(String month)
	   {
	      this.month = month;
	   }
	}